xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+        """
+        self.add(callable)
+        return callable
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+
+        The caller may send() new positional arguments (eg, to implement
+        modifying semantics). Keyword arguments cannot be modified this way.
+        """
+        for handler in self:
```

Comment:
So I had two thoughts on the syntax of validators:

```python
def v(spam, egg):
    return spam == egg

@events.on_spamegg.handler(validator=v)
def se(spam, egg):
    print(spam, egg)
```

```python
@events.on_spamegg.handler
def se(spam, egg):
    print(spam, egg)

@se.validator
def v(spam, egg):
    return spam == egg
```

I prefer the second one myself. I think the implementation will be simpler and leads to a more declarative style.